File: frontend/src/beta/combat/AbilityContracts.js
import { ABILITY_STATS } from "./Abilities.js";
import { MOVE_STATS } from "./Moves.js";

export const EFFECT_TYPES = Object.freeze({
    DAMAGE: "damage",
    HEALING: "healing",
    KNOCKBACK: "knockback",
    PULL: "pull",
    DEBUFF: "debuff",
    INTERRUPT: "interrupt",
    MOVEMENT: "movement",
    TELEPORT: "teleport",
    RESTORE_STATE: "restore_state",
    DAMAGE_REDUCTION: "damage_reduction",
    DAMAGE_IMMUNITY: "damage_immunity",
    DAMAGE_REFLECTION: "damage_reflection",
    SPAWN_ENTITY: "spawn_entity",
});

export const DELIVERY_TYPES = Object.freeze({
    SELF: "self",
    MELEE: "melee",
    RAY: "ray",
    PROJECTILE: "projectile",
    RADIAL: "radial",
    FIELD: "field",
    TRAP: "trap",
    SUMMON: "summon",
    TELEPORT: "teleport",
});

export const SHIELD_MODES = Object.freeze({
    BLOCK: "block",
    IGNORE: "ignore",
    DRAIN_WHILE_ACTIVE: "drain_while_active",
});

export const SHIELD_CHARGE_COSTS = Object.freeze({ ONE: 1, ALL: "all", DISTANCE_SCALED: "distance_scaled" });

const effect = (type, values = {}) => Object.freeze({ type, ...values });
const block = (prevents, values = {}) => Object.freeze({
    mode: SHIELD_MODES.BLOCK,
    halfArcDegrees: 95,
    chargeCost: SHIELD_CHARGE_COSTS.ONE,
    prevents: Object.freeze([...prevents]),
    ...values,
});
const ignore = Object.freeze({ mode: SHIELD_MODES.IGNORE, prevents: Object.freeze([]) });
const drainWhileActive = Object.freeze({ mode: SHIELD_MODES.DRAIN_WHILE_ACTIVE, chargeCost: SHIELD_CHARGE_COSTS.ALL, prevents: Object.freeze([]) });

const A = ABILITY_STATS;
const M = MOVE_STATS;

/**
 * Canonical browser combat metadata. Delivery controls how an effect reaches a
 * target; effects control game-state changes; shieldInteraction filters those
 * effects. Visuals intentionally live outside this catalog.
 */
export const ABILITY_CONTRACTS = Object.freeze({
    swing: contract(DELIVERY_TYPES.MELEE, [effect(EFFECT_TYPES.DAMAGE, { amount: M.swing.damage })], block([EFFECT_TYPES.DAMAGE])),
    block: contract(DELIVERY_TYPES.SELF, [], ignore),
    dash: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.MOVEMENT, { distance: M.dash.distance })], ignore),
    fire_gun: contract(DELIVERY_TYPES.RAY, [effect(EFFECT_TYPES.DAMAGE, { falloff: true })], block([EFFECT_TYPES.DAMAGE])),
    throw_grenade: contract(DELIVERY_TYPES.PROJECTILE, [effect(EFFECT_TYPES.DAMAGE, { falloff: true }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "grenade" })], block([EFFECT_TYPES.DAMAGE], { halfArcDegrees: 180, chargeCost: SHIELD_CHARGE_COSTS.DISTANCE_SCALED })),
    shoot_fireball: contract(DELIVERY_TYPES.PROJECTILE, [effect(EFFECT_TYPES.DAMAGE, { amount: M.shoot_fireball.damage }), effect(EFFECT_TYPES.DEBUFF, { debuff: "burn", durationMs: M.shoot_fireball.burnDurationMs }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "fireball" })], block([EFFECT_TYPES.DAMAGE, EFFECT_TYPES.DEBUFF])),
    stun: contract(DELIVERY_TYPES.MELEE, [effect(EFFECT_TYPES.DAMAGE, { amount: A.stun.damage }), effect(EFFECT_TYPES.DEBUFF, { debuff: "stun", durationMs: A.stun.durationMs })], block([EFFECT_TYPES.DAMAGE, EFFECT_TYPES.DEBUFF])),
    heavy_slash: contract(DELIVERY_TYPES.MELEE, [effect(EFFECT_TYPES.DAMAGE, { amount: A.heavy_slash.damage }), effect(EFFECT_TYPES.DEBUFF, { debuff: "bleed", durationMs: A.heavy_slash.bleedDurationMs })], block([EFFECT_TYPES.DAMAGE, EFFECT_TYPES.DEBUFF], { chargeCost: SHIELD_CHARGE_COSTS.ALL })),
    repulsor_burst: contract(DELIVERY_TYPES.RADIAL, [effect(EFFECT_TYPES.DAMAGE, { amount: A.repulsor_burst.damage }), effect(EFFECT_TYPES.KNOCKBACK, { distance: A.repulsor_burst.knockback })], block([EFFECT_TYPES.DAMAGE])),
    concussive_shot: contract(DELIVERY_TYPES.RAY, [effect(EFFECT_TYPES.DAMAGE, { amount: A.concussive_shot.damage }), effect(EFFECT_TYPES.DEBUFF, { debuff: "slow", durationMs: 2000 })], block([EFFECT_TYPES.DAMAGE, EFFECT_TYPES.DEBUFF])),
    repair_pulse: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.HEALING, { amount: A.repair_pulse.healing })], ignore),
    proximity_mine: contract(DELIVERY_TYPES.TRAP, [effect(EFFECT_TYPES.DAMAGE, { amount: A.proximity_mine.damage }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "proximity_mine" })], block([EFFECT_TYPES.DAMAGE], { halfArcDegrees: 45, chargeCost: SHIELD_CHARGE_COSTS.ALL })),
    quick_jab: contract(DELIVERY_TYPES.MELEE, [effect(EFFECT_TYPES.DAMAGE, { amount: A.quick_jab?.damage ?? M.quick_jab.damage })], block([EFFECT_TYPES.DAMAGE])),
    pistol_shot: contract(DELIVERY_TYPES.RAY, [effect(EFFECT_TYPES.DAMAGE, { falloff: true })], block([EFFECT_TYPES.DAMAGE])),
    rail_shot: contract(DELIVERY_TYPES.RAY, [effect(EFFECT_TYPES.DAMAGE, { amount: A.rail_shot.damage }), effect(EFFECT_TYPES.DEBUFF, { debuff: "shock", durationMs: A.rail_shot.shockDurationMs })], block([EFFECT_TYPES.DAMAGE, EFFECT_TYPES.DEBUFF])),
    gravity_grenade: contract(DELIVERY_TYPES.FIELD, [effect(EFFECT_TYPES.PULL, { perTick: 6 }), effect(EFFECT_TYPES.DAMAGE, { falloff: true }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "gravity_field" })], block([EFFECT_TYPES.DAMAGE], { halfArcDegrees: 45, chargeCost: SHIELD_CHARGE_COSTS.ALL })),
    silence_pulse: contract(DELIVERY_TYPES.PROJECTILE, [effect(EFFECT_TYPES.DEBUFF, { debuff: "silence", durationMs: A.silence_pulse.durationMs }), effect(EFFECT_TYPES.INTERRUPT, { durationMs: A.silence_pulse.interruptMs }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "silence_wave" })], block([EFFECT_TYPES.DEBUFF, EFFECT_TYPES.INTERRUPT])),
    reactive_armor: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.DAMAGE_REDUCTION, { multiplier: 0.5 }), effect(EFFECT_TYPES.DAMAGE_REFLECTION, { multiplier: 0.5 })], ignore),
    hunter_drone: contract(DELIVERY_TYPES.SUMMON, [effect(EFFECT_TYPES.DAMAGE, { amount: A.hunter_drone.damage }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "hunter_drone" })], block([EFFECT_TYPES.DAMAGE])),
    thrust: contract(DELIVERY_TYPES.MELEE, [effect(EFFECT_TYPES.DAMAGE, { amount: M.thrust.damage }), effect(EFFECT_TYPES.KNOCKBACK, { distance: M.thrust.knockback })], ignore),
    micro_dash: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.MOVEMENT, { distance: M.micro_dash.distance })], ignore),
    temporal_rewind: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.RESTORE_STATE, { delayMs: A.temporal_rewind.delayMs }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "temporal_rewind_zone" })], ignore),
    orbital_strike: contract(DELIVERY_TYPES.FIELD, [effect(EFFECT_TYPES.DAMAGE, { amount: A.orbital_strike.damage, falloff: true }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "orbital_zone" })], drainWhileActive),
    absolute_guard: contract(DELIVERY_TYPES.SELF, [effect(EFFECT_TYPES.DAMAGE_IMMUNITY, { durationMs: A.absolute_guard.durationMs })], ignore),
    null_zone: contract(DELIVERY_TYPES.FIELD, [effect(EFFECT_TYPES.DEBUFF, { debuff: "silence", whileInside: true }), effect(EFFECT_TYPES.SPAWN_ENTITY, { entityType: "null_zone" })], ignore),
    phase_strike: contract(DELIVERY_TYPES.TELEPORT, [effect(EFFECT_TYPES.TELEPORT, { passThroughDistance: M.phase_strike.passThroughDistance }), effect(EFFECT_TYPES.DAMAGE, { amount: M.phase_strike.damage })], ignore),
});

function contract(deliveryType, effects, shieldInteraction) {
    return Object.freeze({
        delivery: Object.freeze({ type: deliveryType }),
        effects: Object.freeze(effects),
        shieldInteraction,
    });
}

export function abilityContract(abilityId) {
    return ABILITY_CONTRACTS[abilityId] ?? null;
}

export function hasEffect(abilityId, effectType) {
    return Boolean(abilityContract(abilityId)?.effects.some(({ type }) => type === effectType));
}

File: server/src/main/java/com/example/machiner/simulation/combat/AbilityContracts.java
package com.example.machiner.simulation.combat;

import java.util.List;
import java.util.Map;
import java.util.Set;

/** Gameplay-only ability metadata shared by authoritative combat resolvers. */
public final class AbilityContracts {
    private AbilityContracts() {}

    public enum EffectType { DAMAGE, HEALING, KNOCKBACK, PULL, DEBUFF, INTERRUPT, MOVEMENT, TELEPORT,
        RESTORE_STATE, DAMAGE_REDUCTION, DAMAGE_IMMUNITY, DAMAGE_REFLECTION, SPAWN_ENTITY }
    public enum DeliveryType { SELF, MELEE, RAY, PROJECTILE, RADIAL, FIELD, TRAP, SUMMON, TELEPORT }
    public enum ShieldMode { BLOCK, IGNORE, DRAIN_WHILE_ACTIVE }
    public enum ChargeCost { ONE, ALL, DISTANCE_SCALED }

    public record Effect(EffectType type, String subtype, double amount, int durationMs, boolean runtimeComputed) {
        public Effect(EffectType type) { this(type, null, 0, 0, false); }
    }
    public record ShieldInteraction(ShieldMode mode, double halfArcDegrees, ChargeCost chargeCost,
                                    Set<EffectType> prevents) {
        public boolean prevents(EffectType type) { return prevents.contains(type); }
    }
    public record AbilityContract(DeliveryType delivery, List<Effect> effects, ShieldInteraction shieldInteraction) {}

    private static final ShieldInteraction IGNORE = shield(ShieldMode.IGNORE, 0, ChargeCost.ONE);
    private static final Map<String, AbilityContract> CATALOG = Map.ofEntries(
            entry("swing", DeliveryType.MELEE, block(EffectType.DAMAGE), effect(EffectType.DAMAGE, 20)),
            entry("block", DeliveryType.SELF, IGNORE),
            entry("dash", DeliveryType.SELF, IGNORE, effect(EffectType.MOVEMENT, 400)),
            entry("fire_gun", DeliveryType.RAY, block(EffectType.DAMAGE), computed(EffectType.DAMAGE)),
            entry("throw_grenade", DeliveryType.PROJECTILE, shield(ShieldMode.BLOCK, 180, ChargeCost.DISTANCE_SCALED, EffectType.DAMAGE), computed(EffectType.DAMAGE), spawn("grenade")),
            entry("shoot_fireball", DeliveryType.PROJECTILE, block(EffectType.DAMAGE, EffectType.DEBUFF), effect(EffectType.DAMAGE, 15), debuff("burn", 2, 5000), spawn("fireball")),
            entry("stun", DeliveryType.MELEE, block(EffectType.DAMAGE, EffectType.DEBUFF), effect(EffectType.DAMAGE, 5), debuff("stun", 0, 1200)),
            entry("heavy_slash", DeliveryType.MELEE, shield(ShieldMode.BLOCK, 95, ChargeCost.ALL, EffectType.DAMAGE, EffectType.DEBUFF), effect(EffectType.DAMAGE, 30), debuff("bleed", 2, 5000)),
            entry("repulsor_burst", DeliveryType.RADIAL, block(EffectType.DAMAGE), effect(EffectType.DAMAGE, 20), effect(EffectType.KNOCKBACK, 250)),
            entry("concussive_shot", DeliveryType.RAY, block(EffectType.DAMAGE, EffectType.DEBUFF), effect(EffectType.DAMAGE, 8), debuff("slow", 0, 2000)),
            entry("repair_pulse", DeliveryType.SELF, IGNORE, effect(EffectType.HEALING, 15)),
            entry("proximity_mine", DeliveryType.TRAP, shield(ShieldMode.BLOCK, 45, ChargeCost.ALL, EffectType.DAMAGE), effect(EffectType.DAMAGE, 18), spawn("proximity_mine")),
            entry("quick_jab", DeliveryType.MELEE, block(EffectType.DAMAGE), computed(EffectType.DAMAGE)),
            entry("pistol_shot", DeliveryType.RAY, block(EffectType.DAMAGE), computed(EffectType.DAMAGE)),
            entry("rail_shot", DeliveryType.RAY, block(EffectType.DAMAGE, EffectType.DEBUFF), effect(EffectType.DAMAGE, 40), debuff("shock", 3, 3000)),
            entry("gravity_grenade", DeliveryType.FIELD, shield(ShieldMode.BLOCK, 45, ChargeCost.ALL, EffectType.DAMAGE), effect(EffectType.PULL, 6), computed(EffectType.DAMAGE), spawn("gravity_field")),
            entry("silence_pulse", DeliveryType.PROJECTILE, block(EffectType.DEBUFF, EffectType.INTERRUPT), debuff("silence", 0, 2000), timed(EffectType.INTERRUPT, 100), spawn("silence_wave")),
            entry("reactive_armor", DeliveryType.SELF, IGNORE, effect(EffectType.DAMAGE_REDUCTION, .5), effect(EffectType.DAMAGE_REFLECTION, .5)),
            entry("hunter_drone", DeliveryType.SUMMON, block(EffectType.DAMAGE), effect(EffectType.DAMAGE, 3), spawn("hunter_drone")),
            entry("thrust", DeliveryType.MELEE, IGNORE, effect(EffectType.DAMAGE, 15), effect(EffectType.KNOCKBACK, 30)),
            entry("micro_dash", DeliveryType.SELF, IGNORE, effect(EffectType.MOVEMENT, 150)),
            entry("temporal_rewind", DeliveryType.SELF, IGNORE, timed(EffectType.RESTORE_STATE, 3000), spawn("temporal_rewind_zone")),
            entry("orbital_strike", DeliveryType.FIELD, shield(ShieldMode.DRAIN_WHILE_ACTIVE, 0, ChargeCost.ALL), computed(EffectType.DAMAGE), spawn("orbital_zone")),
            entry("absolute_guard", DeliveryType.SELF, IGNORE, timed(EffectType.DAMAGE_IMMUNITY, 1500)),
            entry("null_zone", DeliveryType.FIELD, IGNORE, debuff("silence", 0, 0), spawn("null_zone")),
            entry("phase_strike", DeliveryType.TELEPORT, IGNORE, effect(EffectType.TELEPORT, 50), effect(EffectType.DAMAGE, 14))
    );

    public static AbilityContract get(String abilityId) {
        AbilityContract contract = CATALOG.get(abilityId);
        if (contract == null) throw new IllegalArgumentException("Unknown ability contract: " + abilityId);
        return contract;
    }

    public static Map<String, AbilityContract> all() { return CATALOG; }

    private static Map.Entry<String, AbilityContract> entry(String id, DeliveryType delivery, ShieldInteraction shield, Effect... effects) {
        return Map.entry(id, new AbilityContract(delivery, List.of(effects), shield));
    }
    private static Effect effect(EffectType type) { return new Effect(type); }
    private static Effect effect(EffectType type, double amount) { return new Effect(type, null, amount, 0, false); }
    private static Effect timed(EffectType type, int durationMs) { return new Effect(type, null, 0, durationMs, false); }
    private static Effect computed(EffectType type) { return new Effect(type, null, 0, 0, true); }
    private static Effect debuff(String subtype, double amount, int durationMs) { return new Effect(EffectType.DEBUFF, subtype, amount, durationMs, false); }
    private static Effect spawn(String entityType) { return new Effect(EffectType.SPAWN_ENTITY, entityType, 0, 0, false); }
    private static ShieldInteraction block(EffectType... prevents) { return shield(ShieldMode.BLOCK, 95, ChargeCost.ONE, prevents); }
    private static ShieldInteraction shield(ShieldMode mode, double arc, ChargeCost cost, EffectType... prevents) {
        return new ShieldInteraction(mode, arc, cost, Set.of(prevents));
    }
}
